### ISSUE TITLE
When nodes are cordoned they are not removed from the target group

### ISSUE THREAD

**Issue by paulalex** [kubernetes-sigs/aws-alb-ingress-controller#1161]
I am running eks 1.14 and deployments with the ingress controller seem to work ok but when I drain a node or cordon a node it stays in the target group meaning its still routable via the ALB.

This has the effect that when running a load test and then draining a node I see 504
 errors when traffic is sent to the node and it is in the process of being terminated.

Is this the desired behaviour or a bug?

**Comment by M00nF1sh** [comment]
Hi, this is not desired behavior, we'll fix this in next release.
currently we relies on EC2's healthy status of nodes.
/kind bug

**Comment by eytan-avisror** [comment]
Check out https://github.com/keikoproj/lifecycle-manager
This removed 504s for us

**Comment by gshutler** [comment]
Is there a rough ETA on the release of v1.1.6? We believe we are also hitting this and if it's in the region of a few weeks we may just wait, but if it's a month plus we will look more deeply at workarounds.

**Comment by eytan-avisror** [comment]
Just wondering how will this be implemented, will the controller reconcile on node events (such as drain)? What happens if the draining node is member of hundreds of target groups? How will clients know until when they should wait post drain? What about terminations that occurs from ASG? There seems to be an inherent race with this approach (what is faster, termination or deregistrstion)

**Comment by M00nF1sh** [comment]
@gshutler the ETA of new release is before Mar 08. (it's a task on my sprint board).
@eytan-avisror Yes, the controller will reconcile on node events and check node's status. And it's unavoidable to update all target groups. 
I didn't get the question of "How will clients know until when they should wait post drain". 
If the termination occurs from ASG, it will guarantee a broken connection.
There is also an unavoidable case for long-live sessions(Websocket), where ALB -> nodeA -> pod(on nodeB), and we cordon nodeA. (there is no way for the pod to initiate proper protocol-specific graceful shutdown).

**Comment by eytan-avisror** [comment]
Thanks @M00nF1sh 
To clarify the question - if a node is getting drained and now starts deregistering from target groups, how can you / whoever is issuing the drain/termination assure it has completed deregistration so that it can safely proceed to terminate?
Imagine an account with 300+ target groups, getting throttled, etc.
So if a node is getting drained / terminated, you only have so much time until it goes down.

**Comment by gshutler** [comment]
Great, thanks @M00nF1sh

**Comment by M00nF1sh** [comment]
@eytan-avisror 
There is no good way to solve this problem as far as i know. one way I can imaging is the controller feedback draining info back to the node via labels/finalizers. But i don't think it's needed once we use the IP mode.
Any ideas?

**Comment by eytan-avisror** [comment]
Definitely not needed with ip mode, however moving to it is not a trivial change and has other implications, I feel that there should be no errors in both modes if we can help it.
We use lifecycle hooks to solve this problem for instance mode (see my first comment about usinf lifecycle-manager), adding a label or publishing an event when work is done will be extremely useful (especially label which is more guaranteed), in this case we could simply use lifecycle hooks to wait for that label instead of hitting AWS APIs so hard in those large scale scenarios

**Comment by M00nF1sh** [comment]
@eytan-avisror 
Yeah, if there is a lot target groups, a fixed wait time don't work for instances.
I think this can be added in V2, where we are plan to periodically check draining status to decided whether remove securityGroups rules into pods/nodes.
I'm wondering whether we can add a finalizer when nodes are added as backend, and remove the finalizer when it's finished draining from targetGroup. (It can be finalizers for each targetGroup or  a single finalizer for all targetGroup).

**Comment by eytan-avisror** [comment]
@M00nF1sh 
You mean the finalizer will be on the node object? It sounds like the right thing to do, but that might be problematic if it means node objects cannot be deleted until they are deregistered, no? What if controller is down and no one is around to remove the finalizer? Nodes will be stuck in deleting if they are terminated? Not sure how cluster will behave with node finalizer during edge cases, but should definitely check.

**Comment by eytan-avisror** [comment]
finalizer on the node also won't help if a termination is already issued, the finalizer is needed on the EC2 instance to be fully safe, and in a sense this is what lifecycle-hooks accomplish since you can keep the instance alive until you are ready for it to terminate.
I think if you simply add a label that says instance is deregistered from all target groups, lifecycle-hooks will be very easy to use in order to have 0 errors on scale-downs, whether they are user initiated or ASG initiated (including AZRebalance).

**Comment by phr3nzii** [comment]
@M00nF1sh any updates on when the next release can roughly be expected?

**Comment by JacobHenner** [comment]
I know this issue was closed by #1172, but I am curious if there are plans to incorporate the label suggestion mentioned here:

>  I think if you simply add a label that says instance is deregistered from all target groups, lifecycle-hooks will be very easy to use in order to have 0 errors on scale-downs, whether they are user initiated or ASG initiated (including AZRebalance).

This would be very useful. My cluster is running with instance targeting, and occasionally the autoscaler will drain and terminate a node before the ingress controller is able to remove the node from all of the ALB target groups. If the controller added this label once target group removal has finished, we'd be able to block premature terminations and prevent this class of 5XXs.

**Comment by Quentin-M** [comment]
PR https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/1172 leads to unexpected behaviors, and in our case - downtime.

Marking a node Unschedulable **does not** mean that the workload on that particular node is unhealthy or should be taken out of service. This is the core difference between Unhealthy and Unschedulable, cordon and drain, etc. Unschedulable simply flags the node to the Kubernetes scheduler in order to exclude it from further scheduling of **new** pods.

As mentioned by another individual in the PR, it is common for operation teams to be marking nodes as Unschedulable when they are reaching EOL and their retirement is desired. Over time, as the pods get updated, they get re-scheduled onto other nodes, so that eventually, said node can be destroyed. However, with this new behavior, taking a node out of the scheduling pool also prompts the ALB controller to take the traffic off the node, and that can happen prior to the new servicing pods being available on schedulable nodes. For instance, one may want to mark unschedulable all of the nodes hosting API servers, and wait for the API team to trigger an API rollout hours/days later which would naturally migrate the workload over to new nodes (e.g. new instance class, etc) -- instead, the aws-alb-controller here just renders the entire API unavailable, whereas the API servers are perfectly healthy and able to service traffic.

Being able to mark nodes Unschedulable independently of the workload and the service traffic is a core element of the separation of concerns principle and administration of base infrastructure independently of the end-user workloads. If the operation teams wants to completely and immediately retire an instance for any reasons (e.g. healthiness concerns), the operation teams would cordon+drain the node instead - the individual pods would then rely on the standard [pod lifecycle](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/) procedure & grace periods for ensuring that traffic is drained properly before the termination of pods.

**Comment by mcpomm** [comment]
I have to agree with Quentin_M. Unschedulable does not mean Unhealthy.
The concept behind cordoning a node is clear. The nodes should simply no longer schedule new workloads.  However, the existing workload on the respective node is still running and available.

I also don't understand why these nodes are taken out of service. Although both the node and the workload are healthy.

We upgraded our EKS cluster to a new version yesterday. Our steps were as follows:

- provisioned a new node group with new image and current EKS version via Terraform.
- adjusted all necessary pdb's and also adjusted the versions of CoreDNS and KubeProxy accordingly......
    After that, we had twice as many nodes in the cluster. Half of them were the new nodes.
- Then we cordoned all the old nodes and wanted to drain them one by one later.
    By cordoning all the old nodes, the evicted workload is immediately distributed to the new nodes when draining.

Funnily enough, we tested this scenario several times on our integration and staging environmen tand there was no such behaviour.
Yesterday's production upgrade from 1.17 to 1.18 unfortunately resulted in downtime because exactly the cordoned nodes were taken out of service although the nodes and workload were still healthy. We had not drained them at that time.

My question: How long does it take until cordoned nodes are taken out of service? And what could be the reason that this behaviour did not occur during previous test upgrades?

My conclusion: Cordoned nodes with a healthy workload and healthy themselves should remain in service. An OPS team must be able to rely on this, otherwise the concept makes no sense at all.
